Interrogator.xUnit/NotConcurrentAttribute.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Interrogator.xUnit.Common;
+using Interrogator.xUnit.Utilities;
+
+namespace Interrogator.xUnit
+{
+	internal static class TypeExtensions
+	{
+		public static IOrderedEnumerable<Type> OrderTypes(this IEnumerable<Type> types)
+			=> types.OrderBy(type => type.FullName);
+	}
+
+	[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+	public class NotConcurrentAttribute : DependsOnAttribute
+	{
+		private static readonly Dictionary<Type, MethodInfo[]> _methodInfoDictionary = new Dictionary<Type, MethodInfo[]>();
+
+		public enum ConcurrencyScope
+		{
+			Class,
+			ClassHierarchy,
+			Namespace,
+			Assembly
+		}
+
+		public string GroupName { get; }
+		public ConcurrencyScope Scope { get; }
+
+		public NotConcurrentAttribute() : this("", ConcurrencyScope.Class) { }
+		public NotConcurrentAttribute(string groupName) : this(groupName, ConcurrencyScope.Class) { }
+
+		public NotConcurrentAttribute(ConcurrencyScope scope) : this("", scope) { }
+
+		public NotConcurrentAttribute(string groupName, ConcurrencyScope scope) : base(null)
+		{
+			GroupName = $"{groupName}_{scope}"; // Adding scope to group name ensures that different scoped attributes don't collide
+			Scope = scope;
+		}
+
+		internal override Result<Option<MethodInfo>, string> TryGetMethod(Type containingType, MemberInfo member, MethodInfo[] testMethods)
+		{
+			var result = (Type ?? containingType)
+				.TryGetMethod(member.Name, ParameterTypes)
+				.Select(currentMethod => GetPreviousInChain(currentMethod, containingType, Scope, GroupName, testMethods));
+			return result;
+		}
+
+		private static Option<MethodInfo> GetPreviousInChain(MethodInfo currentMethod, Type type, ConcurrencyScope scope, string groupName, MethodInfo[] executingMethods)
+		{
+			var methodSet = new HashSet<MemberInfo>(executingMethods);
+			var methodsToExecute = GetSameGroupMethods(scope, type, groupName).Where(methodSet.Contains).ToArray();
+
+			return GetPreviousMethod(currentMethod, methodsToExecute);
+		}
+
+		private static MethodInfo[] GetSameGroupMethods(ConcurrencyScope scope, Type type, string groupName)
+		{
+			if (!_methodInfoDictionary.ContainsKey(type))
+			{
+				IEnumerable<Type> types;
+				switch (scope)
+				{
+					case ConcurrencyScope.Class:
+						types = new[] { type };
+						break;
+					case ConcurrencyScope.Assembly:
+						types = GetTypesInAssembly(type);
+						break;
+					case ConcurrencyScope.Namespace:
+						types = GetTypesInNamespace(type);
+						break;
+					case ConcurrencyScope.ClassHierarchy:
+						types = GetTypesInClassHierarchy(type);
+						break;
+					default:
+						types = Array.Empty<Type>();
+						break;
+				}
+				_methodInfoDictionary[type] = GetMethodsInTypes(groupName, types.ToArray());
```

Comment:
GroupName and scope aren't part of the key, so won't this cause incorrect results depending on which group/scope gets in here first. Can you maybe just prepopulate this entire dictionary in a static constructor with all types that contain methods that are not-concurrent, and then use (type, groupname) as the key. Then the calls never needs to get methods from each type and can instead just get the types for their scope and then index into the dictionary for their type/group. Would that make sense?